File: fastapi/fastapi_book/ch01/gil/run_tasks.go
package main

import (
	"fmt"
	"runtime"
	"sync"
	"time"
)

// CPU 密集型任务：累加到一个很大的数
func cpuIntensiveTask() {
	sum := 0
	for i := range 1000000000 {
		sum += i
	}
}

func run_tasks_sequential(numTasks int) {
	startTime := time.Now()
	for range numTasks {
		cpuIntensiveTask()
	}
	duration := time.Since(startTime)
	fmt.Printf("Go (Sequential) 实验完成，总耗时: %s\n", duration)
}

func run_tasks_concurrent(numTasks int) {
	var wg sync.WaitGroup
	startTime := time.Now()

	// 启动指定数量的 Goroutine
	for range numTasks {
		wg.Add(1) // WaitGroup 计数器加 1
		go func() {
			defer wg.Done() // Goroutine 完成时，计数器减 1
			cpuIntensiveTask()
		}()
	}

	// 等待所有 Goroutine 完成
	wg.Wait()

	duration := time.Since(startTime)
	fmt.Printf("Go (Concurrent) 实验完成，总耗时: %s\n", duration)
}

func main() {
	// 设置要并发执行的任务数量
	numTasks := runtime.NumCPU() // 理想情况下，设置为 CPU 核心数
	fmt.Printf("Go 实验：将在 %d 个 CPU 核心上运行 %d 个并发任务。\n", runtime.NumCPU(), numTasks)

	run_tasks_concurrent(numTasks)
	run_tasks_sequential(numTasks)

}

File: fastapi/fastapi_book/ch01/gil/RunTasks.java
package fastapi_book.ch01.gil;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class RunTasks {

    // CPU-intensive task: sum up to a large number
    private static void cpuIntensiveTask() {
        long sumVal = 0;
        for (long i = 0; i < 1_000_000_000L; i++) {
            sumVal += i;
        }
    }

    // Run tasks sequentially
    private static void runTasksSequential(int numTasks) {
        long startTime = System.currentTimeMillis();

        for (int i = 0; i < numTasks; i++) {
            cpuIntensiveTask();
        }

        long duration = System.currentTimeMillis() - startTime;
        System.out.printf("Java: Sequentially completed %d tasks in: %.2f seconds%n", numTasks, duration / 1000.0);
    }

    // Run tasks with multiple threads
    private static void runTasksThreads(int numTasks) throws InterruptedException {
        ExecutorService executor = Executors.newFixedThreadPool(numTasks);
        long startTime = System.currentTimeMillis();

        // Create and start the specified number of threads
        for (int i = 0; i < numTasks; i++) {
            executor.submit(RunTasks::cpuIntensiveTask);
        }

        // Wait for all threads to complete
        executor.shutdown();
        executor.awaitTermination(1, TimeUnit.MINUTES);

        long duration = System.currentTimeMillis() - startTime;
        System.out.printf("Java: Multi-threaded completion of %d tasks in: %.2f seconds%n", numTasks, duration / 1000.0);
    }

    public static void main(String[] args) throws InterruptedException {
        // Set the number of tasks to run concurrently, ideally equal to the number of CPU cores
        int numTasks = Runtime.getRuntime().availableProcessors();
        System.out.printf("Java (multi-threading) experiment: Running %d tasks on %d CPU cores.%n", numTasks, numTasks);

        runTasksSequential(numTasks);
        runTasksThreads(numTasks);
    }
}
